src/v/raft/state_machine.cc

```diff
@@ -165,23 +167,38 @@ ss::future<result<model::offset>> state_machine::insert_linearizable_barrier(
     /**
      * Inject leader barrier and wait until returned offset is applied
      */
+    if (_barrier_promise) {
+        return _barrier_promise->get_shared_future();
+    }
+    _barrier_promise = barrier_promise_t{};
+    auto f = _barrier_promise->get_shared_future();
+
     return ss::with_timeout(
              timeout,
              _raft->linearizable_barrier().then(
                [this, timeout](result<model::offset> r) {
                    if (!r) {
-                       return ss::make_ready_future<result<model::offset>>(
-                         r.error());
+                       _barrier_promise->set_value(r);
+                       _barrier_promise.reset();
+                       return ss::now();
                    }
 
                    // wait for the returned offset to be applied
-                   return wait(r.value(), timeout).then([r] {
-                       return result<model::offset>(r.value());
+                   return wait(r.value(), timeout).then([this, r] {
```

Comment:
I think there may be some undesirable side effects to this optimization.. In a happy path, the shared future resolves when the wait() succeeds, consider the following sequence of actions..

1. linearizable_barrier at offset o
2. usual replicate() at offset o + 1
3. linearizable_barrier at offset o + 2

(3) can potentially resolve with the same shared future as (1), we have a problem? Normally I don't think this is a problem if we validate the offset returned from linearizable_barrier (3) is  > o + 1 but we don't in many places. Think we have this pattern in topics_frontend.. eg: if we create too many topics back to back and topics stm may not have applied the changes by the time barrier returns.

Reply:
so the barrier semantics was like that before i.e. it could always race with the replicate(), this is why i decided to it this way. This is why we return an offset that is confirmed

Reply:
> so the barrier semantics was like that before 

But  https://github.com/redpanda-data/redpanda/pull/11939/commits/3bcf0388c0de9d44ed0f953e93c19a4e59bfc44b tightens that behavior but this optimization brings it back, no?

> it could always race with the replicate()

just to be clear I'm talking about read-your-own-writes here, so its not just any replicate, its the replicate before the barrier.

Reply:
you are right it may happen that the barrier called after the replicate will return, the value of previous barrier, it won't hurt in this case as that was the previous semantics, but it is definitely confusing. I will drop this commit